src/lib/layout/Searchbar.tsx

```diff
@@ -7,126 +7,205 @@ import {
+  switch (key) {
+    case "ArrowUp":
+      if (current === undefined) return lastIndex;
+      return current <= 0 ? lastIndex : current - 1;
+    case "ArrowDown":
+      if (current === undefined) return 0;
+      return current >= lastIndex ? 0 : current + 1;
+    default:
+      return undefined;
+  }
+};
+
+const Searchbar = () => {
   const [keyword, setKeyword] = useState("");
   const [displayResults, setDisplayResults] = useState(false);
-  const [results, setResults] = useState<SearchResultType[]>([]);
+  const [isTyping, setIsTyping] = useState(false);
+  const [cursor, setCursor] = useState<number>();
 
+  const navigate = useInternalNavigate();
+  const resetHandlerStates = useCallback(() => {
+    setIsTyping(false);
+    setCursor(undefined);
+  }, []);
+  const { results, isLoading } = useSearchHandler(keyword, resetHandlerStates);
   const boxRef = useRef<HTMLDivElement>(null);
-  const chainConfig = getChainConfig();
-
-  let placeholder = "Search by Wallet Address / Tx Hash / ";
-  placeholder += chainConfig.isWasm ? "Code ID / Contract Address" : "Block";
-
-  useEffect(() => {
-    const res: SearchResultType[] = [];
-    if (chainConfig.isWasm && isCodeId(keyword)) res.push("Code ID");
-    else if (isBlock(keyword)) res.push("Block");
-
-    if (!validateContractAddress(keyword)) res.push("Contract Address");
-    if (!validateUserAddress(keyword)) res.push("Wallet Address");
-    if (isTxHash(keyword)) res.push("Transaction Hash");
-    // TODO: add proposal ID
-    setResults(res);
-  }, [
-    chainConfig.isWasm,
-    keyword,
-    validateContractAddress,
-    validateUserAddress,
-  ]);
 
   const handleSearchChange = (e: ChangeEvent<HTMLInputElement>) => {
     const inputValue = e.target.value;
     setKeyword(inputValue);
     setDisplayResults(inputValue.length > 0);
+    setIsTyping(true);
   };
 
-  const handleSelectResult = (type?: SearchResultType, isClick = false) => {
-    AmpTrackUseMainSearch(isClick);
-    const route = type ? getRoute(type) : null;
-    if (route) navigate({ pathname: `${route}/${keyword}` });
-  };
+  const handleSelectResult = useCallback(
+    (type?: SearchResultType, isClick = false) => {
+      AmpTrackUseMainSearch(isClick);
+      const routeOptions = getRouteOptions(type);
+      if (routeOptions) {
+        navigate({
+          pathname: routeOptions.pathname,
+          query: { [routeOptions.query]: keyword },
+        });
+        setDisplayResults(false);
+        setKeyword("");
+      }
+    },
+    [keyword, navigate]
+  );
 
-  const handleOnKeyEnter = (e: KeyboardEvent<HTMLInputElement>) => {
-    if (e.key === "Enter") handleSelectResult(results.at(0));
-  };
+  const handleOnKeyEnter = useCallback(
+    (e: KeyboardEvent<HTMLInputElement>) => {
+      if (!results.length) return;
+      const lastIndex = results.length - 1;
+      const nextCursor = getNextCursor(e.key, cursor, lastIndex);
+      const listItem = document.getElementById(`item-${nextCursor}`);
```

Comment:
probably cleaner to just create a function that handles the next cursor then set the value within it self and just call the function in the switch case. With that, these variables would not be needed